app/models/g5_authenticatable/user.rb

```diff
@@ -47,6 +47,23 @@ def update_roles_from_auth(auth_data)
       end
     end
 
+    def clients
+      return G5Updatable::Client.all if has_global_role
+      G5Updatable::Client.where(id: client_roles.map(&:resource_id))
+    end
+
+    def client_roles
+      G5Authenticatable::Role.joins('INNER JOIN g5_updatable_clients ON g5_updatable_clients.id = g5_authenticatable_roles.resource_id')
+        .where('g5_authenticatable_roles.resource_type = ?', G5Updatable::Client.name)
+    end
+
+    def has_global_role
+      has_role?(:admin) ||
```

Comment:
I like making it a policy in g5_authenticable. Not all apps use g5_header
but many apps might need to know which clients a user has access to.

On Thu, Jun 4, 2015 at 11:16 AM, Maeve Revels <notifications@github.com>
wrote:

> In app/models/g5_authenticatable/user.rb
> <https://github.com/G5/g5_authenticatable/pull/37#discussion_r31748735>:
>
> > @@ -47,6 +47,23 @@ def update_roles_from_auth(auth_data)
> >        end
> >      end
> >
> > +    def clients
> > +      return G5Updatable::Client.all if has_global_role
> > +      G5Updatable::Client.where(id: client_roles.map(&:resource_id))
> > +    end
> > +
> > +    def client_roles
> > +      G5Authenticatable::Role.joins('INNER JOIN g5_updatable_clients ON g5_updatable_clients.id = g5_authenticatable_roles.resource_id')
> > +        .where('g5_authenticatable_roles.resource_type = ?', G5Updatable::Client.name)
> > +    end
> > +
> > +    def has_global_role
> > +      has_role?(:admin) ||
>
> @jonnysamps <https://github.com/jonnysamps>, @vgarro
> <https://github.com/vgarro> Yes, I agree. Depending on how this method is
> intended to be used, I think has_global_role probably should be a helper
> method on the base policy. Then the client app could define its own custom
> policies to restrict access to its own models based on global vs scoped
> roles.
>
> I'm guessing that we will need to define a new pundit policy that
> restricts access to clients. In particular, in order to populate the
> dropdown menu of clients in the header, we need a custom scope that will
> look up the clients associated with a user based on roles (similar to the
> G5Updatable::ClientPolicy
> <https://github.com/g5search/g5-authentication/blob/sprint/app/policies/g5_updatable/client_policy.rb>
> on the server side, although the actual implementation will differ). I'm
> not sure whether this new policy should be part of g5_authenticatable
> itself, or if it's something that should go into g5_header (I'm leaning
> towards the latter, but I'm open to suggestions).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/G5/g5_authenticatable/pull/37/files#r31748735>.
>



-- 
Jonathan